HLTrigger/Muon/interface/HLTTriMuonIsolation.h

```diff
@@ -0,0 +1,242 @@
+    iEvent.getByToken(L3DiMuonsFilterToken_, L3DiMuonsFilterCands);
+    
+    std::vector<reco::RecoChargedCandidateRef> PassedL3Muons;
+    L3DiMuonsFilterCands->getObjects(trigger::TriggerMuon, PassedL3Muons);
+  
+    // Get the Trk + L3 muon candidates (after merging)
+    edm::Handle<reco::RecoChargedCandidateCollection> AllMuCands;
+    iEvent.getByToken(AllMuonsToken_, AllMuCands);
+    
+    // Get iso tracks
+    edm::Handle<reco::TrackCollection> IsoTracks;
+    iEvent.getByToken(IsoTracksToken_, IsoTracks);
+          
+    if (AllMuCands->size() >= 3 && L3MuCands->size() >= 2){
+        // Create the 3-muon candidates
+        // loop over L3/Trk muons and create all combinations
+        auto AllMuCands_end = AllMuCands->end();
+        for (auto i = AllMuCands->begin(); i != AllMuCands_end; ++i) {
+            // check that muon_i passes the previous filter
+            bool passingPreviousFilter_1 = false;
+            const reco::TrackRef &tk_i = i->track();
+            for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                if (reco::deltaR2(tk_i->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_1 = true;
+            }
+            for (auto j = i+1; j != AllMuCands_end; ++j) {
+                // check that muon_j passes the previous filter
+                bool passingPreviousFilter_2 = false;
+                const reco::TrackRef &tk_j = j->track();
+                for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                    reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                    if (reco::deltaR2(tk_j->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_2 = true;                
+                }
+                for (auto k = j+1; k != AllMuCands_end; ++k){
+                    // check that muon_k passes the previous filter
+                    bool passingPreviousFilter_3 = false;
+                    const reco::TrackRef &tk_k = k->track();
+                    for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                        reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                        if (reco::deltaR2(tk_k->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_3 = true;
+                    }                    
+                    // at least two muons must have passed the previous di-muon filter
+                    if (!( (passingPreviousFilter_1 & passingPreviousFilter_2 ) ||
+                           (passingPreviousFilter_1 & passingPreviousFilter_3 ) ||
+                           (passingPreviousFilter_2 & passingPreviousFilter_3 ) )) continue;
+
```

Comment:
You mean the same muons appears twice in the collection?
Otherwise, the nested loops start at +1 index off the outer loop index, so the same
muon is not used twice.